File: module_3/python/longest_palindrome.py
"""
Title: Longest palindrome

Problem:
    Given a string s, find the longest palindromic substring in s. You may
    assume that the maximum length of s is 1000.

Execution: python longest_palindrome.py
"""
import unittest


def longest_palindrome(s: str):
    """Function for calculating longest palindrome."""
    res = ""
    curr_length = 0
    for i in range(len(s)):
        if is_palindrome(s, i-curr_length-1, i):
            res = s[i-curr_length-1:i+1]
            curr_length = curr_length + 2
        elif is_palindrome(s, i-curr_length, i):
            res = s[i-curr_length:i+1]
            curr_length = curr_length + 1
    return res


def is_palindrome(s: str, begin: int, end: int):
    """Helper function to check if string is palindrome."""
    if begin < 0:
        return False
    while begin < end:
        if s[begin] != s[end]:
            return False
        begin += 1
        end -= 1
    return True


class TestLongestPalindrome(unittest.TestCase):
    """Unit test for longest palindrome."""

    def test_1(self):
        """Test for longest palindrome."""
        test_palindrome = "babad"
        self.assertEqual(longest_palindrome(test_palindrome), "bab")
        print("Explanation: 'aba' is also a valid answer.")

    def test_2(self):
        """Test for longest palindrome."""
        test_palindrome = "cbbd"
        self.assertEqual(longest_palindrome(test_palindrome), "bb")


if __name__ == '__main__':
    unittest.main()

File: module_3/java/LongestPalindrome.java
/*
 *   Title: Longest palindrome
 *
 *   Problem:
 *
 *   Given a string s, find the longest palindromic substring in s. You
 *   may assume that the maximum length of s is 1000.
 *
 *   Execution: javac LongestPalindrome.java && java LongestPalindrome
 */


public class LongestPalindrome {
    public static String longestPalindrome(String s) {
        String res = "";
        int currLength = 0;
        for(int i=0; i<s.length(); i++){
            if(isPalindrome(s, i-currLength-1, i)) {
                res = s.substring(i-currLength-1, i+1);
                currLength = currLength + 2;
            }
            else if(isPalindrome(s, i-currLength, i)) {
                res = s.substring(i-currLength, i+1);
                currLength = currLength + 1;
            }
        }
        return res;
    }

    public static boolean isPalindrome(String s, int begin, int end){
        if(begin < 0) {
            return false;
        }
        while(begin < end) {
        	if(s.charAt(begin++) != s.charAt(end--)) { 
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        assert longestPalindrome("babad") == "bab";
        System.out.println("Note: 'aba' is also a valid answer");

        assert longestPalindrome("cbbd") == "bb";

        System.out.println("Passed all test cases");
    }
    
}
